nomad/node_endpoint.go

```diff
@@ -1150,6 +1152,26 @@ func (n *Node) UpdateAlloc(args *structs.AllocUpdateRequest, reply *structs.Gene
 		return fmt.Errorf("must update at least one allocation")
 	}
 
+	// Ensure the node is allowed to update allocs.
+	// The node needs to successfully heartbeat before updating its allocs.
+	nodeID := args.NodeID
+	if nodeID == "" {
+		// COMPAT 1.4
+		// Maintain backwards compatibility with clients that don't set the
+		// NodeID field in the request.
+		nodeID = args.Alloc[0].NodeID
+	}
```

Comment:
I tried to use `n.ctx.NodeID` but you're right, it's not present when the request is forwarded. I think it would take quite a bit of refactoring since we would need to pass it in the `n.src.forward` call and that affects all endpoints.

Another option would be to add it to [`InternalRPCInfo`](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/structs/structs.go#L228-L233) before forwarding the request, but I'm not sure if that's a good 😅

Reply:
I think if you just grab it before calling `forward` it would work, but that's a bit odd as forward is almost always the first thing we do.

What if we just dropped the field and always plucked it off the alloc? That would avoid adding another only-used-in-1-of-3-places field to the overloaded request struct. We already always have it on the stripped allocs, so we'd finally make that decision sensible.

We can ship it with the new NodeID field as its not worth holding up these excellent fixes for.

Reply:
> I think if you just grab it before calling `forward` it would work

The problem with this is that the leader would then need to skip the validation, or more specifically, a server would need to ignore the check if the request was forward and assume that the check has been done at some point. 

I thought this would be a bigger problem for cross-region request since the first server would not find the node, but now that I think about it, it's kind of weird to have a cross-region request in this endpoint 😅 

> What if we just dropped the field and always plucked it off the alloc? 

Yeah, that sounds good to me. I wasn't sure which way to go any way and this is probably the most reliable option without adding more code 👍